File: Courses/edX/HarvardX/CS50's Introduction to Computer Science/Week 1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int luhn(int a[]);
int n = 0;

int main(void)
{
    // prompt users for input
    long int i = get_long("Number: ");

    long int y = i;

    // calculate number of digits
    while (y > 0)
    {
        y = y / 10;
        n++;
    }

    // storing digit in an array
    y = i;
    int a[n];
    for (int j = 0; j < n; j++)
    {
        a[j] = y % 10;
        y = y / 10;
    }

    // checking card using luhn
    if (luhn(a) == 1)
    {
        // checking type of card
        y = i;
        for (int j = 0; j < (n - 2); j++)
        {
            y = y / 10;
        }
        // checking amex
        if (n == 15 && (y == 34 || y == 37))
        {
            printf("AMEX\n");
        }
        // checking mastercard
        else if (n == 16 && (y >= 51  && y <= 55))
        {
            printf("MASTERCARD\n");
        }
        // checking visa
        else if ((n == 13 || n == 16) && (y >= 40  && y <= 49))
        {
            printf("VISA\n");
        }
        // if nothing then invalid
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

int luhn(int a[])
{
    // making digits at even places twice
    for (int j = 1; j < n; j = j + 2)
    {
        a[j] = a[j] * 2;
    }

    // adding sum of digits at even places and digits at odd places
    int sum = 0;
    for (int j = 0; j < n; j++)
    {
        sum = sum + (a[j] % 10);
        a[j] = a[j] / 10;
    }

    for (int j = 0; j < n; j++)
    {
        sum = sum + (a[j] % 10);
        a[j] = a[j] / 10;
    }

    // returning result
    if (sum % 10 == 0)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

File: Courses/edX/HarvardX/CS50's Introduction to Computer Science/Week 6/pset6/sentimental-credit/credit.py
# TODO
from cs50 import get_string
from sys import exit
import re

# Prompt User Input
card = get_string("Number: ")

# Amex 15 Digits, Starts with - 34 or 37
# Visa 13 or 16 Digits, Starts with - 4
# Master 16 Digits, Starts with - 51 to 55

# Calculating Digits
digits = len(card)

# Validating Basic Conditions
if ((digits not in [13, 15, 16]) or ((card[0] != '4') and (card[0] != '3' and card[1] not in ['4', '7']) and (card[0] != '5' and card[1] not in ['1', '2', '3', '4', '5']))):
    print("INVALID")
    exit(1)

rev = card[::-1]
sum = 0

# Luhn Algorithm
for i in range(digits):
    if (i % 2 == 0):
        sum = sum + int(rev[i])
    else:
        sum = sum + ((int(rev[i]) * 2) % 10)
for i in range(digits):
    if (i % 2 == 1):
        sum = sum + (((int(rev[i]) * 2) // 10) % 10)

# Print Result
if (sum % 10 == 0):
    if (card[0] == '4'):
        print("VISA")
    elif (card[0] == '3' and card[1] in ['4', '7']):
        print("AMEX")
    else:
        print("MASTERCARD")
else:
    print("INVALID")
